default/python/universe_generation/monsters.py

```diff
@@ -1,8 +1,82 @@
 import random
 import freeorion as fo
 import util
+from util import MapGenerationError
 import statistics
 import universe_tables
+from galaxy import DisjointSets
+
+
+class StarlaneAlteringMonsters(object):
+    def __init__(self, systems):
+        self.systems = systems
+        self.placed = set()
+        self.removed_lanes = set()
+        self.starlanes = [(s1, s2) for s1 in systems for s2 in fo.sys_get_starlanes(s1) if s1 < s2]
+
+    def can_place_at(self, system, plan):
+        # Check if the starlane altering monster fleet ''plan'' can be placed at ''system''
+        # without disjoining the galaxy map.
+        # Compute the disjoint set of the galaxy without the starlanes
+        # from the proposed system.  Return False if there will be more
+        # connected regions than the number of placed starlane altering
+        # monsters plus one, otherwise True.
+        local_lanes = {(min(system, s), max(system, s)) for s in fo.sys_get_starlanes(system)}
+
+        dsets = DisjointSets()
+        for s3 in self.systems:
+            dsets.add(s3)
+
+        for lane in self.starlanes:
+            if lane in self.removed_lanes or lane in local_lanes:
+                continue
+            dsets.link(lane[0], lane[1])
+
+        num_contiguous_regions = len(dsets.complete_sets())
+        expected_num_contiguous_regions = (len(self.placed) + 2)
+
+        if num_contiguous_regions > expected_num_contiguous_regions:
+            return False
+
+        if num_contiguous_regions < expected_num_contiguous_regions:
+            util.report_error("Number of contiguous regions %d is below the expected number "
+                              "of contiguous regions %d when placing %d monster %s that can "
+                              "break starlanes."
+                              % (num_contiguous_regions, expected_num_contiguous_regions,
+                                 len(self.placed) + 1, plan.name()))
+            return False
+
+        return True
+
+    def place(self, system, plan):
+        # create map altering monster fleet ''plan'' at ''system''.
+        local_lanes = {(min(system, s), max(system, s)) for s in fo.sys_get_starlanes(system)}
+
+        populate_monster_fleet(plan, system)
+        if system not in self.placed:
+            self.placed.add(system)
+            self.removed_lanes |= local_lanes
+
+
+def populate_monster_fleet(fleet_plan, system):
+    """
+    Create a monster fleet in ''system'' according to ''fleet_plan''
+    """
+
+    # create monster fleet
+    monster_fleet = fo.create_monster_fleet(system)
+
+    if monster_fleet == fo.invalid_object():
+        errstr = "Python generate_monsters: unable to create new monster fleet %s" % fleet_plan.name()
+        raise MapGenerationError(errstr)
```

Comment:
Think about that from other side. 

You just bring new approach for this projects. You create first custom exception. Make first handler of custom exception.

For me this approach is ok, but I don't think that we should bring it now.  You will not save any lines on it.